src/twisted/internet/test/test_tcp.py

```diff
@@ -1230,12 +1230,26 @@ def stopReactorIfServerAccepted(_):
 
     serverConnectionMade.addCallback(stopReactorIfServerAccepted)
 
-    port = listen(reactor, serverFactory)
-    listeningHost = port.getHost()
     clientFactory = MyClientFactory()
-    connect(reactor, listeningHost, clientFactory)
 
-    reactor.callWhenRunning(exhauster.exhaust)
+    if hasattr(reactor, "callLater"):
+        # For Glib-based reactors, the exhauster should be run after the signal handler used by glib [1]
+        # and restoration of twisted signal handlers [2], so I use callLater with non-zero time
+        # [1] https://gitlab.gnome.org/GNOME/pygobject/-/blob/3.42.0/gi/_ossighelper.py#L76
+        # [2] https://github.com/twisted/twisted/blob/twisted-22.4.0/src/twisted/internet/_glibbase.py#L134
```

Comment:
I'd probably argue that `_glibbase.py` is violating the `IReactorCore` contract by calling `callWhenRunning` functions before it has gotten its signal handlers in place. :/

I don't really want to ask you to fix that for this PR (the problems looks gnarly to me, depending on how three different implementations of `_handleSignals` interact, at least) ... however I also think it is true that it sucks that any application code that really wants to be safe will have to duplicate this `callLater` logic instead of using `callWhenRunning`.  Fortunately, it's a really obscure cases where `callWhenRunning` breaks the interface, I guess.

I'd be okay with a compromise of describing this misbehavior of `_glibbase.py` in a new ticket so there's at least a slightly greater chance of someone fixing the underlying issue.  It would probably be good to link to that ticket from here and vice versa.

Reply:
Yeah that sounds unusual. I don't think I can fix this in the short term, so I created a ticket as a reminder: https://twistedmatrix.com/trac/ticket/10342